src/shared/tpm2-util.c

```diff
@@ -3797,45 +3819,85 @@ char *tpm2_pcr_mask_to_string(uint32_t mask) {
         return TAKE_PTR(s);
 }
 
-int tpm2_pcr_mask_from_string(const char *arg, uint32_t *ret_mask) {
-        uint32_t mask = 0;
+int hex_to_bytes(uint8_t bytes[SHA256_DIGEST_SIZE], const char *hex) {
+        for (int i=0, j=0; i<SHA256_DIGEST_SIZE*2; i+=2, j++) {
+                uint8_t f1 = (hex[i] % 32 + 9) % 25;
+                uint8_t f2 = (hex[i+1] % 32 + 9) % 25;
+                if (f1 >= 16 || f2 >= 16 ) {
+                        return -EINVAL;
+                }
+                bytes[j] = f1 * 16 + f2;
+        }
+        return 0;
+}
+
+int tpm2_pcr_from_string(const char *arg, uint32_t *ret_mask, uint32_t *ret_literal_mask, uint8_t ret_literal[][SHA256_DIGEST_SIZE]) {
+        uint32_t mask = 0, literal_mask=0;
         int r;
 
         assert(arg);
         assert(ret_mask);
+        assert(ret_literal);
 
         if (isempty(arg)) {
                 *ret_mask = 0;
+                *ret_literal_mask = 0;
                 return 0;
         }
 
         /* Parses a "," or "+" separated list of PCR indexes. We support "," since this is a list after all,
          * and most other tools expect comma separated PCR specifications. We also support "+" since in
          * /etc/crypttab the "," is already used to separate options, hence a different separator is nice to
-         * avoid escaping. */
+         * avoid escaping.
+         * In case there are literals specified on the argument, they are expected to have the format
+         * <index>:sha256=<hash>*/
 
         const char *p = arg;
         for (;;) {
                 _cleanup_free_ char *pcr = NULL;
                 unsigned n;
+                char *r2;
 
                 r = extract_first_word(&p, &pcr, ",+", EXTRACT_DONT_COALESCE_SEPARATORS);
                 if (r == 0)
                         break;
                 if (r < 0)
                         return log_error_errno(r, "Failed to parse PCR list: %s", arg);
 
-                r = pcr_index_from_string(pcr);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to parse specified PCR or specified PCR is out of range: %s", pcr);
-                n = r;
-                SET_BIT(mask, n);
+                /* detect if this is (potentially) a pcr index or a pcr literal*/
+                r2 = strchr(pcr, ':');
+
+                /* if this is a pcr literal, try to parse it */
+                if (r2) {
+                        unsigned pcr_idx;
+                        char pcr_hex[2*SHA256_DIGEST_SIZE+1];
+                        r = sscanf(pcr, "%u:sha256=%s", &pcr_idx, pcr_hex);
```

Comment:
this should be reasonably agnostic to hash functions. i.e. do not hardcode sha256 here. 

also, please don't parse with sscanf(). Use extract_first_word(), or or startswith() or so.